runtime/src/iree/hal/drivers/hip/hip_device.c

```diff
@@ -935,6 +1012,47 @@ static iree_status_t iree_hal_hip_device_queue_alloca(
     iree_hal_buffer_t** IREE_RESTRICT out_buffer) {
   iree_hal_hip_device_t* device = iree_hal_hip_device_cast(base_device);
 
+  iree_status_t status = iree_ok_status();
+  if (device->supports_memory_pools &&
+      !iree_all_bits_set(params.type, IREE_HAL_MEMORY_TYPE_HOST_VISIBLE)) {
+    iree_hal_buffer_t* buffer = NULL;
+
+    IREE_RETURN_IF_ERROR(iree_hal_hip_memory_pools_prepare_buffer(
+        &device->memory_pools, device->hip_dispatch_stream, pool, params,
+        allocation_size, &buffer));
+
+    iree_status_t status = iree_hal_deferred_work_queue_enqueue_alloc(
+        device->work_queue, wait_semaphore_list, signal_semaphore_list, buffer);
+    if (!iree_status_is_ok(status)) {
+      iree_hal_hip_buffer_allocation_failed(buffer);
+      iree_hal_resource_release(&buffer->resource);
+      return status;
+    }
+    *out_buffer = buffer;
+    return iree_ok_status();
+  } else if (!iree_all_bits_set(params.type,
+                                IREE_HAL_MEMORY_TYPE_HOST_VISIBLE) &&
+             iree_hal_hip_allocator_isa(
+                 iree_hal_device_allocator(base_device))) {
+    iree_hal_buffer_t* buffer = NULL;
+
+    IREE_RETURN_IF_ERROR(iree_hal_hip_device_pepare_async_alloc(
+        device, params, allocation_size, &buffer));
+
+    iree_status_t status = iree_hal_deferred_work_queue_enqueue_alloc(
+        device->work_queue, wait_semaphore_list, signal_semaphore_list, buffer);
+    if (iree_status_is_ok(status)) {
+      status = iree_hal_deferred_work_queue_issue(device->work_queue);
+    }
+    if (!iree_status_is_ok(status)) {
+      iree_hal_hip_buffer_allocation_failed(buffer);
+      iree_hal_resource_release(&buffer->resource);
+      return status;
```

Comment:
as above, don't do multiple returns - you're returning status 2 lines below and this can instead follow the return convention we use in the codebase:
```c
if (iree_status_is_ok(status)) {
  *out_buffer = buffer;
} else {
  iree_hal_hip_buffer_allocation_failed(buffer);
  iree_hal_resource_release(&buffer->resource);
}
return status;
```

This avoids ambiguity (you'll never set out_buffer unless status is ok, and if you don't assign the out_buffer you are guaranteed to have cleaned up after yourself).